torch/_inductor/simple_fsdp/reorder.py

```diff
@@ -68,53 +18,76 @@ def reorder_all_gather(
     result_list: List[scheduler.BaseSchedulerNode] = []
     all_gather_list: List[scheduler.BaseSchedulerNode] = []
     node_to_type: Dict[scheduler.BaseSchedulerNode, int] = {}
-
     inverse_users, node_users = compute_node_users(snodes)
-    snodes.reverse()
 
     for node in snodes:
         node_to_type[node] = get_node_type(node)
 
+    snodes.reverse()
     for idx, node in enumerate(snodes):
         node_type = node_to_type[node]
-        if node_type in [NodeType.REDUCE_SCATTER, NodeType.COMPUTE]:
+        if node_type in [NodeType.REDUCE_SCATTER, NodeType.COMPUTE, NodeType.RS_WAIT]:
             # we do not reorder reduce scatter and compute node
             if node not in result_list and node not in all_gather_list:
                 result_list.append(node)
         elif node_type == NodeType.ALL_GATHER:
             # gather i-th all gather node and its dependencies
             all_gather_list.append(node)
             inverse_user = list(inverse_users[node])
-            if len(inverse_user) > 0:
+            if (
+                len(inverse_user) > 0
+                and not node_to_type[inverse_user[0]] == NodeType.ALL_GATHER
+                and inverse_user not in all_gather_list
+            ):
                 all_gather_list.extend(inverse_user)
-        elif node_type == NodeType.WAIT:
+        elif node_type == NodeType.AG_WAIT:
             if (
-                node_to_type[snodes[idx + 1]] == NodeType.ALL_GATHER
+                (
+                    (
+                        node_to_type[snodes[idx + 1]] == NodeType.AG_WAIT
+                        and node_to_type[snodes[idx + 2]] == NodeType.ALL_GATHER
+                    )
+                    or (
+                        node_to_type[snodes[idx + 1]] == NodeType.ALL_GATHER
+                        and node_to_type[snodes[idx - 1]] != NodeType.AG_WAIT
+                    )
+                )
```

Comment:
I think you are trying to cover the cases with/without bucketing, but I can't parse this. Let's discuss in person before I make any suggestions.